Add scheduler benchmark tests for PodAntiAffinity rules

misterikkit opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This adds a benchmark test that measures the scheduling latency of pods with PodAntiAffinity rules where topologyKey=kubernetes.io/hostname.

Additionally, this PR imposes a minimum `b.N` value to make the test results more meaningful.

The benchmark is needed to validate any performance improvements made for #54189.


Command used to invoke:
```sh
make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_ARGS="-run=xxxx -bench=."
```

Sample output:
```
pkg: k8s.io/kubernetes/test/integration/scheduler_perf
BenchmarkScheduling/100Nodes/0Pods-40        100          36316838 ns/op
BenchmarkScheduling/100Nodes/1000Pods-40                     100          69460001 ns/op
BenchmarkScheduling/1000Nodes/0Pods-40                       100         193755560 ns/op
BenchmarkScheduling/1000Nodes/1000Pods-40                    100         343451472 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/250Pods-40                  250         127236420 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-40                 250         478661925 ns/op
PASS
ok      k8s.io/kubernetes/test/integration/scheduler_perf       354.008s
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
Add scheduler benchmark tests for affinity rules.

This adds new benchmark tests that measure scheduler latency of pods
that use affinity rules. Specifically, this tests affinity rules with
topologyKey="kubernetes.io/hostname".

misterikkit commented:
/sig scheduling
/assign bsalamat

bsalamat commented on test/integration/scheduler_perf/scheduler_bench_test.go:
```diff
@@ -17,44 +17,87 @@ limitations under the License.
+		{nodes: 500, pods: 400},
+	}
+	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc")
+	// NewSimpleWithControllerCreatePodStrategy will add name=rc to pods.
```
Please put the comment above the function call.

misterikkit replied:
done

bsalamat commented on test/integration/scheduler_perf/scheduler_bench_test.go:
```diff
@@ -17,44 +17,87 @@ limitations under the License.
+	basePod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "affinity-pod-",
+			Labels:       map[string]string{"testType": "affinity"},
```
Would it be better to have matching labels in our (anti)-affinity rule?

misterikkit replied:
I wasn't sure here. My thought was that all the pods scheduled during benchmark have anti-affinity to the pre-scheduled pods, but not each other. Do you think the pods should also have anti-affinity with themselves?

The testType label isn't serving any purpose other than possibly debugging. Maybe just omit it?

bsalamat replied:
I don't see any config that specifies anti-affinity to existing pods. Given that the number of pods in your tests is smaller than or equal to the number of nodes, I think it would make the tests better if we added anti-affinity among the created pods.

misterikkit replied:
Per our discussion, I've added a lower bound to b.N (and rebased on top of that commit). I also updated the tests' PodAntiAffinity rules in the second fixup (27d1ce3).

Long term, I think we will end up moving away from micro-benchmarks for this sort of test.

bsalamat reviewed (commented):
Thanks, @misterikkit! Looks good. Just a couple of minor comments.

Commit:
Set a minimum b.N for scheduler_perf benchmarks.

Under the default behavior of Go benchmark tests, all our scheduler_perf
benchmark tests run with b.N=1, which is lower than we would like. This
adds a lower bound to b.N so that the results are more meaningful.

The alternative to this change is to always run these tests with the
-benchtime flag set to a duration which will force b.N to increase. That
would cause any test setup to be executed repeatedly as b.N ramps up,
and -timout would probably also need to be set higher.

misterikkit commented:
I have updated the AntiAffinity benchmark per our (further) discussion. It now,
- sets up on X pods without any affinity rules
- schedules b.N pods with PodAntiAffinity against themselves.

This lets us set X to a value much larger than the number of nodes, creating a more realistic test.

bsalamat commented on test/integration/scheduler_perf/scheduler_bench_test.go:
```diff
@@ -31,22 +34,80 @@ import (
-		{nodes: 1000, pods: 0},
-		{nodes: 1000, pods: 1000},
+	tests := []struct{ nodes, pods, minOps int }{
+		{nodes: 100, pods: 0, minOps: 100},
```
Please rename `pods` to `existingPods` and `minOps` to `minPods`. Also, please add a comment explaining that existingPods are created before the measurements start and then at least `minPods` are created.

misterikkit replied:
Done

bsalamat commented:
Thanks, @misterikkit. Looks good. Please go ahead and squash your commits.

misterikkit commented:
Squashed.

bsalamat commented:
/lgtm

bsalamat commented:
please run hack/update-bazel.sh in your git directory.

Commit:
Auto generated BUILD file

misterikkit commented:
fixed build file

bsalamat commented:
/lgtm

bsalamat commented:
/ok-to-test
